WalletWasabi/Blockchain/TransactionBuilding/SmartCoinSelector.cs

```diff
@@ -17,22 +17,38 @@ public SmartCoinSelector(List<SmartCoin> unspentCoins)
 	}
 
 	private List<SmartCoin> UnspentCoins { get; }
+	private int IterationCount { get; set; }
 
-	/// <param name="unused">Unused parameter, make it an empty list.</param>
-	public IEnumerable<ICoin> Select(IEnumerable<ICoin> unused, IMoney target)
+	/// <param name="suggestion">We use this to detect if NBitcoin tries to suggest something different and indicate the error.</param>
+	public IEnumerable<ICoin> Select(IEnumerable<ICoin> suggestion, IMoney target)
 	{
-		var targetMoney = target as Money;
+		var targetMoney = (Money)target;
 
 		long available = UnspentCoins.Sum(x => x.Amount);
 		if (available < targetMoney)
 		{
 			throw new InsufficientBalanceException(targetMoney, available);
 		}
 
+		// This is hacky safety check to make sure the first few iteration can go through.
+		if (IterationCount > 10)
+		{
+			var suggestedSum = suggestion.Sum(c => (Money)c.Amount);
+			if (suggestedSum < (Money)target)
+			{
+				throw new InsufficientBalanceException(targetMoney, suggestedSum);
+			}
+		}
+
+		if (IterationCount > 500)
+		{
+			throw new TimeoutException("Coin selection timed out.");
```

Comment:
Here the returned value doesn't depend on the `suggested` parameter and then, in the specific scenario that we are trying to fix, the method returns the exact same thing again and again until it throws.

Why not to use the suggested coin set instead?

Reply:
I was trying to mimic the exact same behavior as before but except for the specific case. If we are getting back to the original solution this commit would not work anymore and it was added for a reason: 
https://github.com/zkSNACKs/WalletWasabi/commit/9251d08be684e053b58c0626464e8c56f9410ac1